### PR TITLE
Adding craft doc for RLP

### PR DESCRIPTION
**Pull request by Boomatang**
Add a more detail guide to for the sections of a RLP.

## Verification
As the change is to the docs and the main mode of render will be the docs.kuadrant.io there is a draft PR opened to the docs site which includes these changes. That PR can be found here https://github.com/Kuadrant/docs.kuadrant.io/pull/39

Expected:
* Guide explains all components in the RLP.
* Link in the guides are crafted for use on the site and may not work as expected if the Markdown file is view in isolation.

### CODE REVIEW

File: doc/craft/ratelimitpolicy.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
+# Crafting A RateLimitPolicy
+A detail explanation of the fields in the RateLimitPolicy is outlined here.
+For a quick reference refer to [Reference](/kuadrant-operator/doc/reference/ratelimitpolicy/) section.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:3]: Link is broken in GitHub.

File: doc/craft/ratelimitpolicy.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+```yaml
+apiVersion: kuadrant.io/v1beta2
+kind: RatelimitPplicy
+```
+
+## Metadata
+When creating the RateLimitPolicy the policy must be crated in the same namespace as the resource being targeted.
+```yaml
+metadata:
+    name: toystore
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:17]: Any special reason to be using 4-space indentation? I think most (if not all) our examples use 2 spaces.

File: doc/craft/ratelimitpolicy.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+## apiVersion and kind
+The latest version of the RateLimitPolicy is `kuadrant.io/v1beta2`.
+With the kind being `RateLimitPolicy`.
+```yaml
+apiVersion: kuadrant.io/v1beta2
+kind: RatelimitPplicy
+```
+
+## Metadata
+When creating the RateLimitPolicy the policy must be crated in the same namespace as the resource being targeted.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:14]:
```suggestion
The RateLimitPolicy must be created in the same namespace as the resource being targeted.
```

File: doc/craft/ratelimitpolicy.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+```yaml
+metadata:
+    name: toystore
+    namespace: default
+```
+Once created Kuadrant may add and manage additional annotations and labels.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:20]:
```suggestion
Once created, Kuadrant may add additional annotations and labels to the policy.
```

File: doc/craft/ratelimitpolicy.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+```yaml
+metadata:
+    annotations:
+        kuadrant.io/namespace: default
+```
+
+## Spec
+
+### limits
+`spec.limits` is a list of named limits, which describe what limits should be placed on a resource.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:31]: What does "resource" mean here?

File: doc/craft/ratelimitpolicy.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+spec:
+    limits:
+        limit_1:
+            ...
+        limit_2:
+            ...
+```
+
+#### counters
+`counters` defines additional rate limit counters based on context qualifiers and well known selectors.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:44]: I wouldn't say the counters are "additional." These are the qualifiers that define _the_ counters that are created for each possible value assumed for the dynamic attributes specified here.

File: doc/craft/ratelimitpolicy.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+The full list of well known attributes can be found [here](/architecture/rfcs/0002-well-known-attributes/#request-attributes).
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:45]: Broken link.

Also, pointing to one group of well-known attributes only (i.e. request attributes.) Attributes from the other groups also apply.

File: doc/craft/ratelimitpolicy.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+```yaml
+spec: 
+    limits:
+        limit_1:
+            counters:
+                - request.url_path
+```
+The above example creates a new counter for every url_path that is requested.
+
+A more concrete example of why creating new counters base off the well known attributes may relate to comments on social network.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:56]: Sorry. I don't understand this sentence.

File: doc/craft/ratelimitpolicy.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                    - 'api.toystore.com'
+                    - 'admin.toystore.com'
+                    - '*.example.com'
+                  matches:
+                    - method: GET
+                      path:
+                        type: PathPrefix
+                        value: /
+```
+The above example selects an number of hostnames to protect, and the `matches` refines the selection to GET requests on any URI that is prefix with `/`.
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:130]:
```suggestion
The above example selects a number of hostnames to protect, and the `matches` refines the selection to GET requests on any URI that is prefix with `/`.
```

File: doc/craft/ratelimitpolicy.md [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    limits:
+        limit_1:
+            when:
+                - operator: "eq"
+                  selector: "request.useragent"
+                  value: "curl/8.0.1"
+```
+The above example will match HTTP requests that are sent with the header HTTP_USER_AGENT set with a value of "curl/8.0.1". 
+
+### targetRef
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:193]: Maybe explain this before explaining `routeSelectors`?

File: doc/craft/ratelimitpolicy.md [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+metadata:
+    name: toystore
+    namespace: default
+spec: 
+    limits:
+        toystore:
+            rates:
+                - duration: 10
+                  limit: 5
+                  unit: secound
```
Comment (guicassolato) [inline doc/craft/ratelimitpolicy.md:223]:
```suggestion
                  unit: second
```

**Review by guicassolato — CHANGES_REQUESTED** [review]
I understand this is intended to be some RLP 101. However, IMHO, it overloads a bit with https://docs.kuadrant.io/kuadrant-operator/doc/rate-limiting; it adds new wording for fields and concepts more carefully defined in the [Reference](https://docs.kuadrant.io/kuadrant-operator/doc/reference/ratelimitpolicy/), as well as in https://docs.kuadrant.io/kuadrant-operator/doc/reference/route-selectors/.

I think bits of this doc could perhaps be extracted to enhance other existing docs – either the ones I linked here or some of the more basic user guides. But, honestly, I don't think we need it as-is as part of our documentation.

Not that the approach is not desirable. On the contrary! I'm sure first-time users could really use such a detailed explanation that slices down each block and level of the RLP. Maybe it could be turned into a blog post instead, with a context use case to drive the explanations?

File: doc/craft/ratelimitpolicy.md (hunk 2 above)
Comment (Boomatang) [reply doc/craft/ratelimitpolicy.md:17]: No special reason, my editor has a default of 4 spaces. I always found it as a good guide to highlight deeply nested code block and the cyclomatic complexity that brings.

File: doc/craft/ratelimitpolicy.md (hunk 10 above)
Comment (Boomatang) [reply doc/craft/ratelimitpolicy.md:193]: I explained the different sections in an alphabetically order as this is how a user would see items listed in a CR.  If this was to be ordered by the perceived importances should `routeSelectors` be listed before `counters`?

**Comment by Boomatang** [comment]
@guicassolato and I had a good conversation about this change. We are in agreement that the information contained in the proposed doc is valuable but adding this information by this doc might not be the best way. The next step is to review this content with the aim of refactoring it into the existing documents (overview, reference and how-to guides).

Marking this PR as a draft for now.

**Pull request closed without merging**